src/gpu/intel/jit/ngen/ngen_elf.hpp

```diff
@@ -358,7 +531,7 @@ std::vector<uint8_t> ELFCodeGenerator<hw>::getBinary(const std::vector<uint8_t>
 
     binary.resize(paddedSzELF + paddedSzMetadata + paddedSzKernel);
 
-    (void) new(binary.data()) ZebinELF(paddedSzKernelName, metadata.size(), kernel.size());
+    (void) new(binary.data()) ZebinELF(paddedSzKernelName, metadata.size(), kernel.size(), __LINE__);
```

Comment:
The debug information should come from the implementation. In particular, it will need passed into [make_kernel](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/2da2a65552814a455b0c2735ae24a94d39c9c6de/src/gpu/intel/jit/jit_generator.hpp#L188) and forwarded down to this function call from locations like: [gen_pooling](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/2da2a65552814a455b0c2735ae24a94d39c9c6de/src/gpu/intel/jit/pooling/gen_pooling.cpp#L150-L151), [gen_conv](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/2da2a65552814a455b0c2735ae24a94d39c9c6de/src/gpu/intel/jit/conv/gen_convolution.cpp#L206-L246), [gen_reorder](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/2da2a65552814a455b0c2735ae24a94d39c9c6de/src/gpu/intel/jit/reorder/gen_reorder.cpp#L208-L209), and also into the [gen_gemm_kernels calls to getBinary](https://github.com/intel-innersource/libraries.performance.math.onednn/blob/2da2a65552814a455b0c2735ae24a94d39c9c6de/src/gpu/intel/jit/gemm/gen_gemm_kernel.cpp#L895-L914)

Reply:
I am suggesting the latter option, that the "location responsible for building the kernel" is set further up the stack. The core issue with the current method is that very different implementations (such as GEMM and pooling) will point to the same source code location. By pointing to the implementation generator, the debug info will be much more relevant.

> So are you suggesting that we add more comprehensive debug information from higher up the stack? If so, what would you include?

I think this is a good idea, for example, I am thinking we should add an asm->source code line mapping so that debugging is simpler within our team. This likely requires invasive changes, so I expect I will be handling it. Is there is any additional information that would be useful to collect from your perspective?

Reply:
Okay that's clear, thanks. More information is always better than less. Our tool in its current form doesn't do anything that would require more debug info than what I've added here, but I can certainly imagine enhancements that indicate different properties of the kernel in the flame graph visualization. For these we might need to package some more parameters with the debug info such as the data type.

I also agree that mapping instructions back to source lines would be very useful. Our tool currently shows stall counts for individual kernel instructions, so it would be nice to add functionality that let's you zoom in on the kernels source and see a visualization of EU stalls per line.